core/consensus/common/src/select_chain.rs

```diff
@@ -0,0 +1,54 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate Consensus Common.
+
+// Substrate Demo is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate Consensus Common is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate Consensus Common.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::error::Error;
+use runtime_primitives::traits::{Block as BlockT, NumberFor};
+
+
+/// The SelectChain trait defines the strategy upon which the head is chosen
+/// if multiple forks are present for an opaque definition of "best" in the 
+/// specific chain build.
+////
+/// The Strategy can be customised for the two use cases of authoring new blocks
+/// upon the best chain or which fork to finalise. Unless implemented differently
+/// by default finalisation methods fall back to use authoring, so as a minimum
+/// `_authoring`-functions must be implemented. 
+///
+/// Any particular user must make explicit, however, whether they intend to finalise
+/// or author through the using the right function call, as these might differ in
+/// some implementations.
+///
+/// Non-deterministicly finalising chains may only use the `_authoring` functions.
+pub trait SelectChain<Block: BlockT>: Sync + Send + Clone {
+
+	/// Get all leaves of the chain: block hashes that have no children currently.
+	/// Leaves that can never be finalized will not be returned.
+	fn leaves(&self) -> Result<Vec<<Block as BlockT>::Hash>, Error>;
+
+	/// Among those `leaves` deterministically pick one chain as the generally
+	/// best chain to author new blocks upon and probably finalize.
+	fn best_chain(&self) -> Result<<Block as BlockT>::Header, Error>;
+
+	/// Get the most recent block hash of the best chain that contain block
+	/// with the given `target_hash` for finalisation
```

Comment:
This is a break from what we discussed while designing, which was that `finality_target` would find an _ancestor_ of `target_hash`. So to find a block to finalize, you do `finality_target(best_chain())`. For a lot of chain selection systems, `finality_target` is the identity function, but for some we might not be able to finalize all the blocks in that chain, so it would look backwards a bit.

Reply:
I must have mixed something up then, because I thought this was the API we discussed. I think I just copied the docs over and didn't update them properly. But the idea was that given `best_chain()` (which is what most finality gadget will use prior) the SC would pick the most recent block that can be/should be finalized.

I'll update the doc to reflect that.

Reply:
@rphmeier I changed the docs to reflect closer what we had in mind. Please confirm, that this the API we had discussed.

Reply:
@gnunicorn This is a good point. I think the only solution is to specify that the `finality_target` is the _block we would like to see finalized_, which excludes blocks that are already finalized. So a `finality_target` should be able to return `None` to indicate that there is no viable chain to finalize.